File: apps/api/services/cleaning.py
"""
CivicMind AI — Data Cleaning Service

Implements automated data quality operations: duplicate removal, missing value
detection and imputation, outlier detection and soft-capping, date normalization,
and column profiling with statistical summaries.
"""

import math
from typing import Any


class CleaningReport:
    """Container for cleaning pipeline statistics."""

    def __init__(self):
        self.before_rows: int = 0
        self.after_rows: int = 0
        self.missing_values_count: int = 0
        self.duplicates_removed: int = 0
        self.outliers_detected: int = 0
        self.normalized_dates: int = 0
        self.cleaned_rows: list[dict[str, Any]] = []

    def to_dict(self) -> dict:
        return {
            "before_rows": self.before_rows,
            "after_rows": self.after_rows,
            "missing_values_count": self.missing_values_count,
            "duplicates_removed": self.duplicates_removed,
            "outliers_detected": self.outliers_detected,
            "normalized_dates": self.normalized_dates,
        }


def clean_dataset(
    rows: list[dict[str, Any]],
    schema: list[dict[str, str]],
) -> CleaningReport:
    """Run the full cleaning pipeline on a list of row dictionaries.

    Steps:
    1. Deduplicate rows (by non-id fields)
    2. Detect and impute missing values
    3. Normalize date fields to ISO 8601
    4. Detect and soft-cap numeric outliers

    Args:
        rows: The raw data rows to clean.
        schema: Column metadata list with ``name`` and ``type`` keys.

    Returns:
        A ``CleaningReport`` containing statistics and cleaned rows.
    """
    report = CleaningReport()
    report.before_rows = len(rows)

    # ── Step 1: Deduplicate ─────────────────────────────────────────────
    unique_rows: list[dict[str, Any]] = []
    seen: set[str] = set()

    for row in rows:
        # Build a fingerprint excluding the 'id' field
        key_parts = {k: v for k, v in row.items() if k != "id"}
        fingerprint = str(sorted(key_parts.items()))
        if fingerprint in seen:
            report.duplicates_removed += 1
        else:
            seen.add(fingerprint)
            unique_rows.append(dict(row))

    # ── Step 2–4: Per-row cleaning ──────────────────────────────────────
    cleaned: list[dict[str, Any]] = []
    for row in unique_rows:
        new_row = dict(row)
        for field in schema:
            col_name = field["name"]
            col_type = field["type"]
            val = new_row.get(col_name)

            # Missing value detection
            is_missing = (
                val is None
                or val == ""
                or (isinstance(val, float) and math.isnan(val))
            )

            if is_missing:
                report.missing_values_count += 1
                # Impute with sensible defaults
                if col_type == "number":
                    new_row[col_name] = 0
                elif col_type == "string":
                    new_row[col_name] = "Unknown"
                elif col_type == "boolean":
                    new_row[col_name] = False
                elif col_type == "date":
                    from datetime import datetime, timezone

                    new_row[col_name] = datetime.now(timezone.utc).isoformat()
            else:
                # Date normalization
                if col_type == "date":
                    try:
                        from datetime import datetime

                        parsed = datetime.fromisoformat(str(val).replace("Z", "+00:00"))
                        iso_str = parsed.isoformat()
                        if str(val) != iso_str:
                            new_row[col_name] = iso_str
                            report.normalized_dates += 1
                    except (ValueError, TypeError):
                        from datetime import datetime, timezone

                        new_row[col_name] = datetime.now(timezone.utc).isoformat()
                        report.normalized_dates += 1

                # Outlier detection (domain-aware thresholds)
                if col_type == "number" and isinstance(val, (int, float)):
                    capped = _detect_and_cap_outlier(col_name, val)
                    if capped is not None:
                        report.outliers_detected += 1
                        new_row[col_name] = capped

        cleaned.append(new_row)

    report.after_rows = len(cleaned)
    report.cleaned_rows = cleaned
    return report


def _detect_and_cap_outlier(col_name: str, value: float) -> float | None:
    """Apply domain-specific outlier capping rules.

    Returns the capped value if an outlier was detected, or None otherwise.
    """
    rules = [
        # Emergency response time > 2 hours → cap to 45 mins
        ("time", lambda v: v > 120, 45),
        # AQI outside 0–500 → cap to 150
        ("aqi", lambda v: v < 0 or v > 500, 150),
        # Percentage rates > 150 → cap to 100
        ("rate", lambda v: v < 0 or v > 150, 100),
        # Transit delay > 600s → cap to 120s
        ("delay", lambda v: v > 600, 120),
    ]

    for keyword, check_fn, cap_value in rules:
        if keyword in col_name.lower() and check_fn(value):
            return cap_value

    return None


def profile_columns(
    rows: list[dict[str, Any]],
    schema: list[dict[str, str]],
) -> list[dict[str, Any]]:
    """Generate a statistical profile for each column in the dataset.

    Args:
        rows: The dataset rows.
        schema: Column metadata.

    Returns:
        A list of column profile dictionaries containing fill rates,
        min/max/avg for numerics, and unique counts for strings.
    """
    profiles: list[dict[str, Any]] = []

    for field in schema:
        col_name = field["name"]
        col_type = field["type"]
        values = [
            r[col_name]
            for r in rows
            if r.get(col_name) is not None and r[col_name] != ""
        ]
        total = len(rows)
        filled = len(values)
        fill_rate = round((filled / total) * 100, 1) if total > 0 else 0.0

        stats: dict[str, Any] = {}

        if col_type == "number":
            nums = [v for v in values if isinstance(v, (int, float))]
            if nums:
                stats["min"] = min(nums)
                stats["max"] = max(nums)
                stats["avg"] = round(sum(nums) / len(nums), 2)

        elif col_type == "string":
            str_vals = [str(v) for v in values]
            freq: dict[str, int] = {}
            for sv in str_vals:
                freq[sv] = freq.get(sv, 0) + 1
            top_values = sorted(freq.items(), key=lambda x: -x[1])[:5]
            stats["unique_count"] = len(freq)
            stats["top_values"] = [
                {"value": v, "count": c} for v, c in top_values
            ]

        profiles.append({
            "column_name": col_name,
            "type": col_type,
            "fill_rate": fill_rate,
            "stats": stats,
        })

    return profiles

File: server/cleaning.ts
export interface CleaningReport {
  beforeRows: number;
  afterRows: number;
  missingValuesCount: number;
  duplicatesRemoved: number;
  outliersDetected: number;
  normalizedDates: number;
  cleanedRows: any[];
}

export function cleanDataset(rows: any[], schema: any[]): CleaningReport {
  const beforeRows = rows.length;
  let missingValuesCount = 0;
  let duplicatesRemoved = 0;
  let outliersDetected = 0;
  let normalizedDates = 0;

  // 1. Remove duplicate rows based on string representation of everything except 'id'
  const uniqueRows: any[] = [];
  const seen = new Set<string>();

  for (const row of rows) {
    const { id, ...rest } = row;
    const str = JSON.stringify(rest);
    if (seen.has(str)) {
      duplicatesRemoved++;
    } else {
      seen.add(str);
      uniqueRows.push({ ...row });
    }
  }

  // 2. Perform cleaning, missing value detection, outlier detection and date normalization on unique rows
  const cleanedRows = uniqueRows.map((row, index) => {
    const newRow = { ...row };

    for (const field of schema) {
      const val = newRow[field.name];

      // Missing value check
      if (val === undefined || val === null || val === "" || (typeof val === "number" && isNaN(val))) {
        missingValuesCount++;
        // Apply smart defaults based on type
        if (field.type === "number") {
          newRow[field.name] = 0; // standard numeric default
        } else if (field.type === "string") {
          newRow[field.name] = "Unknown";
        } else if (field.type === "boolean") {
          newRow[field.name] = false;
        } else if (field.type === "date") {
          newRow[field.name] = new Date().toISOString();
        }
      } else {
        // Date Normalization
        if (field.type === "date") {
          try {
            const parsedDate = new Date(val);
            if (!isNaN(parsedDate.getTime())) {
              const originalStr = String(val);
              const isoStr = parsedDate.toISOString();
              if (originalStr !== isoStr) {
                newRow[field.name] = isoStr;
                normalizedDates++;
              }
            } else {
              newRow[field.name] = new Date().toISOString();
              normalizedDates++;
            }
          } catch {
            newRow[field.name] = new Date().toISOString();
            normalizedDates++;
          }
        }

        // Outlier detection (simple statistical boundary for known numeric parameters)
        if (field.type === "number" && typeof val === "number") {
          // Detect unreasonable numbers based on municipal operations context
          if (
            (field.name.includes("time") && val > 120) || // emergency response taking > 2 hours
            (field.name.includes("aqi") && (val < 0 || val > 500)) || // invalid air quality index
            (field.name.includes("rate") && (val < 0 || val > 150)) || // percentage-like rate out of bounds
            (field.name.includes("delay") && val > 600) // transit delays > 10 mins
          ) {
            outliersDetected++;
            // Soft-bound outlier cap instead of deleting row
            if (field.name.includes("time") && val > 120) newRow[field.name] = 45; // cap to 45 mins
            if (field.name.includes("aqi") && val > 500) newRow[field.name] = 150;
            if (field.name.includes("rate") && val > 150) newRow[field.name] = 100;
          }
        }
      }
    }
    return newRow;
  });

  return {
    beforeRows,
    afterRows: cleanedRows.length,
    missingValuesCount,
    duplicatesRemoved,
    outliersDetected,
    normalizedDates,
    cleanedRows,
  };
}

export function profileColumns(rows: any[], schema: any[]) {
  return schema.map((field) => {
    const values = rows.map((r) => r[field.name]).filter((v) => v !== undefined && v !== null && v !== "");
    const totalCount = rows.length;
    const filledCount = values.length;
    const fillRate = totalCount > 0 ? (filledCount / totalCount) * 100 : 0;

    let stats: any = {};
    if (field.type === "number") {
      const nums = values.filter((v) => typeof v === "number") as number[];
      if (nums.length > 0) {
        const min = Math.min(...nums);
        const max = Math.max(...nums);
        const sum = nums.reduce((a, b) => a + b, 0);
        const avg = sum / nums.length;
        stats = { min, max, avg: parseFloat(avg.toFixed(2)) };
      }
    } else if (field.type === "string") {
      const stringVals = values.map(String);
      const frequencies: { [key: string]: number } = {};
      stringVals.forEach((v) => {
        frequencies[v] = (frequencies[v] || 0) + 1;
      });
      const topValues = Object.entries(frequencies)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 5)
        .map(([value, count]) => ({ value, count }));
      stats = { uniqueCount: Object.keys(frequencies).length, topValues };
    }

    return {
      columnName: field.name,
      type: field.type,
      fillRate: parseFloat(fillRate.toFixed(1)),
      stats,
    };
  });
}
